routing_algorithms/map_matching.hpp

```diff
@@ -0,0 +1,627 @@
+            }
+        }
+
+        void initialize(const Matching::CandidateLists& candidates_list)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            osrm::json::Array states;
+            for (unsigned t = 0; t < candidates_list.size(); t++)
+            {
+                osrm::json::Array timestamps;
+                for (unsigned s = 0; s < candidates_list[t].size(); s++)
+                {
+                    osrm::json::Object state;
+                    state.values["transitions"] = osrm::json::Array();
+                    state.values["coordinate"] = osrm::json::make_array(
+                        candidates_list[t][s].first.location.lat / COORDINATE_PRECISION,
+                        candidates_list[t][s].first.location.lon / COORDINATE_PRECISION);
+                    state.values["viterbi"] = osrm::json::clamp_float(Matching::IMPOSSIBLE_LOG_PROB);
+                    state.values["pruned"] = 0u;
+                    timestamps.values.push_back(state);
+                }
+                states.values.push_back(timestamps);
+            }
+            osrm::json::get(*object, "states") = states;
+        }
+
+        void add_transition_info(const unsigned prev_t,
+                                 const unsigned current_t,
+                                 const unsigned prev_state,
+                                 const unsigned current_state,
+                                 const double prev_viterbi,
+                                 const double emission_pr,
+                                 const double transition_pr,
+                                 const double network_distance,
+                                 const double great_circle_distance)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            osrm::json::Object transistion;
+            transistion.values["to"] = osrm::json::make_array(current_t, current_state);
+            transistion.values["properties"] =
+                osrm::json::make_array(osrm::json::clamp_float(prev_viterbi),
+                                       osrm::json::clamp_float(emission_pr),
+                                       osrm::json::clamp_float(transition_pr),
+                                       network_distance,
+                                       great_circle_distance);
+
+            osrm::json::get(*object, "states", prev_t, prev_state, "transitions")
+                .get<mapbox::util::recursive_wrapper<osrm::json::Array>>()
+                .get().values.push_back(transistion);
+
+        }
+
+        void set_viterbi(const std::vector<std::vector<double>>& viterbi,
+                         const std::vector<std::vector<bool>>& pruned)
+        {
+            // json logger not enabled
+            if (!logger)
+                return;
+
+            for (auto t = 0u; t < viterbi.size(); t++)
```

Comment:
- a range-based for loop with an irange would allow to make `t` const
- `t++` should be `++t`